rust/private/clippy.bzl

```
@@ -0,0 +1,167 @@
+        type = crate_type,
+        root = crate_root_src(ctx, srcs = rust_srcs),
+        srcs = rust_srcs,
+        deps = ctx.rule.attr.deps,
+        proc_macro_deps = ctx.rule.attr.proc_macro_deps,
+        aliases = ctx.rule.attr.aliases,
+        output = None,
+        edition = get_edition(ctx.rule.attr, toolchain),
+        rustc_env = ctx.rule.attr.rustc_env,
+    )
+
+    dep_info, build_info = collect_deps(
+        ctx.label,
+        crate_info.deps,
+        crate_info.proc_macro_deps,
+        crate_info.aliases,
+        toolchain,
+    )
+
+    compile_inputs, out_dir = collect_inputs(
+        ctx,
+        toolchain,
+        crate_info,
+        dep_info,
+        build_info
+    )
+
+    args, env = construct_arguments(
+        ctx,
+        toolchain,
+        crate_info,
+        dep_info,
+        output_hash = None,
+        rust_flags = [])
+
+    # A marker file indicating clang tidy has executed successfully.
+    # This file is necessary because "ctx.actions.run" mandates an output.
+    clippy_marker = ctx.actions.declare_file(ctx.label.name + "_clippy.ok")
+
+    command = construct_compile_command(
+        ctx,
+        toolchain.clippy_driver.path,
+        toolchain,
+        crate_info,
+        build_info,
+        out_dir,
+    ) + (" && touch %s" % clippy_marker.path)
+
+    # Deny the default-on clippy warning levels.
+    #
+    # If these are left as warnings, then Bazel will consider the execution
+    # result of the aspect to be "success", and Clippy won't be re-triggered
+    # unless the source file is modified.
+    args.add("-Dclippy::style")
+    args.add("-Dclippy::correctness");
+    args.add("-Dclippy::complexity");
+    args.add("-Dclippy::perf");
+
+    ctx.actions.run_shell(
+        command = command,
+        inputs = compile_inputs,
+        outputs = [clippy_marker],
+        env = env,
+        tools = [toolchain.clippy_driver],
+        arguments = [args],
+        mnemonic = "Clippy",
+    )
+
+    return [
+        OutputGroupInfo(clippy_checks = depset([clippy_marker])),
+    ]
+
+# Executes a "compile command" on all specified targets.
+#
+# Example: Run the clang-tidy checker on all targets in the codebase.
+#   bazel build --aspects=@rules_rust//rust:clippy.bzl%rust_clippy_aspect \
+#               --output_groups=clippy_checks \
+#               //...
+rust_clippy_aspect = aspect(
+    attr_aspects = ["deps"],
+    fragments = ["cpp"],
+    host_fragments = ["cpp"],
+    attrs = {
+        "_cc_toolchain": attr.label(
+            default = Label("@bazel_tools//tools/cpp:current_cc_toolchain"),
+        ),
+    },
+    toolchains = [
+        "@io_bazel_rules_rust//rust:toolchain",
+        "@bazel_tools//tools/cpp:toolchain_type"
+    ],
+    implementation = _clippy_aspect_impl,
+)
+
+def _rust_clippy_rule_impl(ctx):
+    checks = depset([])
+    for dep in ctx.attr.deps:
+        checks = depset(dep[OutputGroupInfo].clippy_checks.to_list(), transitive = checks.to_list())
```

Overusing `to_list` leads to quadratic behavior. It is better to leverage the depset structure:

```suggestion
        checks = depset(transitive = [dep[OutputGroupInfo], ctx.attr.deps])
```

I am not sure the above works, if not you can simply extract the list:
```suggestion
        checks = depset(transitive = [dep[OutputGroupInfo]] + ctx.attr.deps)
```

---

The first option throws an error:
`at index 0 of transitive, got element of type OutputGroupInfo, want depset`

The second option also throws the same error:
`at index 0 of transitive, got element of type OutputGroupInfo, want depset`.

---

Sorry that was meant to be `dep[OutputGroupInfo].clippy_checks` where I put `dep[OutputGroupInfo]`

---

Attempting:
`checks = depset(transitive = [dep[OutputGroupInfo].clippy_checks, ctx.attr.deps])`
Causes the error:
`at index 0 of transitive, got element of type list, want depset`

Attempting:
`checks = depset(transitive = [dep[OutputGroupInfo].clippy_checks] + ctx.attr.deps)`
Causes the error:
`at index 0 of transitive, got element of type Target, want depset`

---

Ok sorry for making you try several thing. From the error the correct way should be:

```
depset(transitive = ctx.attr.deps, direct = dep[OutputGroupInfo].clippy_checks)
```

---

Using:

```
checks = depset([])
for dep in ctx.attr.deps:
  checks = depset(transitive = ctx.attr.deps, direct = dep[OutputGroupInfo].clippy_checks)
return [DefaultInfo(files = checks)]
```

Results in:

```
File "/home/smklein/repos/rules_rust/rust/private/clippy.bzl", line 156, in _rust_clippy_rule_impl
                depset(<2 more arguments>)
for direct, got depset, want sequence
```

I read through https://docs.bazel.build/versions/master/skylark/depsets.html, and it seems like the following works:

```
files = depset([], transitive = [dep[OutputGroupInfo].clippy_checks for dep in ctx.attr.deps])
return [DefaultInfo(files = files)]
```

---

Oh right sorry I got all mixed up. Yes the solution you gave is the correct one, it avoids creating N depset

---

Updated